x/aswap/handler.go

```diff
@@ -0,0 +1,292 @@
+	if err != nil {
+		return nil, err
+	}
+
+	swapAddr := weave.NewAddress(swap.PreimageHash)
+	amount, err := h.bank.Balance(db, swapAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	// withdraw the money from swap to recipient
+	if err := moveCoins(db, h.bank, swapAddr, swap.Recipient, amount); err != nil {
+		return nil, err
+	}
+
+	// Delete swap when empty.
+	if err := h.bucket.Delete(db, swap.PreimageHash); err != nil {
+		return nil, err
+	}
+	return &weave.DeliverResult{}, nil
+}
+
+// validate does all common pre-processing between Check and Deliver.
+func (h ReleaseSwapHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*Swap, error) {
+	var msg ReleaseSwapMsg
+	if err := weave.LoadMsg(tx, &msg); err != nil {
+		return nil, errors.Wrap(err, "load msg")
+	}
+
+	swap, err := loadSwap(h.bucket, db, hashPreimage(msg.Preimage))
+	if err != nil {
+		return nil, err
+	}
+
+	// Sender must authorize this.
+	if !h.auth.HasAddress(ctx, swap.Src) {
+		return nil, errors.ErrUnauthorized
+	}
+
+	if IsExpired(ctx, swap.Timeout) {
+		err := errors.Wrapf(errors.ErrExpired, "swap expired %v", swap.Timeout)
+		return nil, err
+	}
+
+	return swap, nil
+}
+
+// ReturnSwapHandler returns funds to the sender when swap timed out.
+type ReturnSwapHandler struct {
+	auth   x.Authenticator
+	bucket Bucket
+	bank   cash.Controller
+}
+
+var _ weave.Handler = ReturnSwapHandler{}
+
+// Check just verifies it is properly formed and returns
+// the cost of executing it.
+func (h ReturnSwapHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.CheckResult, error) {
+	_, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	return &weave.CheckResult{GasAllocated: returnSwapCost}, nil
+}
+
+// Deliver moves all the tokens from the swap to the defined sender if
+// all preconditions are met. The swap is deleted afterwards.
+func (h ReturnSwapHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*weave.DeliverResult, error) {
+	swap, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return nil, err
+	}
+
+	swapAddr := weave.NewAddress(swap.PreimageHash)
+	available, err := h.bank.Balance(db, swapAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	// withdraw all coins from swap to the defined "sender"
+	if err := moveCoins(db, h.bank, swapAddr, swap.Src, available); err != nil {
+		return nil, err
+	}
+	if err := h.bucket.Delete(db, swap.PreimageHash); err != nil {
+		return nil, err
+	}
+	return &weave.DeliverResult{}, nil
+}
+
+// validate does all common pre-processing between Check and Deliver.
+func (h ReturnSwapHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*Swap, error) {
+	var msg ReturnSwapMsg
+	if err := weave.LoadMsg(tx, &msg); err != nil {
+		return nil, errors.Wrap(err, "load msg")
+	}
+
+	swap, err := loadSwap(h.bucket, db, msg.PreimageHash)
+	if err != nil {
+		return nil, err
+	}
+
+	// Sender must authorize this.
+	if !h.auth.HasAddress(ctx, swap.Src) {
```

Comment:
req: It is in the best interest for the original swap creator to send this message but I don't think we should limit this to him

Reply:
Yes, I was thinking about this, I had a todo to solve it. I think the idea here is anyone should be able to do this, WDYT?